openassessment/workflow_batch_update_api.py

```diff
@@ -374,99 +357,98 @@ def get_workflow_update_parameters(peer_workflows):
-
-            # retrieve course block
-            course_block_key = UsageKey.from_string(peer_workflow.course_id)
-            course_block = store.get_item(course_block_key)
-
-            if is_flexible_peer_grading_on(ora_block, course_block):
-                # we are interested only in submissions for ORA with flexible peer grading configured
-                course_settings = {
-                    'force_on_flexible_peer_openassessments': course_block.force_on_flexible_peer_openassessments}
-                _add_workflow_update_parameters(workflow_update_parameters_dict,
-                                                peer_workflow.course_id,
-                                                peer_workflow.item_id,
-                                                peer_workflow.submission_uuid,
-                                                ora_block.workflow_requirements(),
-                                                course_settings)
+            if peer_workflow.course_id not in cache:
+                # retrieve course block
+                course_block_key = UsageKey.from_string(peer_workflow.course_id)
+                course_block = store.get_item(course_block_key)
+
+                # retrieve openassessment block
+                ora_block_key = UsageKey.from_string(peer_workflow.item_id)
+                ora_block = store.get_item(ora_block_key)
+
+                if is_flexible_peer_grading_on(ora_block, course_block):
+                    # we are interested only in submissions for ORA with flexible peer grading configured
+                    course_settings = {
+                        'force_on_flexible_peer_openassessments': course_block.force_on_flexible_peer_openassessments}
+
+                    workflow_update_data = _add_course_data(
+                        workflow_update_data=workflow_update_data,
+                        course_id=peer_workflow.course_id,
+                        item_id=peer_workflow.item_id,
+                        submission_uuid=peer_workflow.submission_uuid,
+                        assessment_requirements=ora_block.workflow_requirements(),
+                        course_settings=course_settings)
+
+                cache.add(peer_workflow.course_id)
+            elif peer_workflow.item_id not in cache:
+                # retrieve openassessment block
+                ora_block_key = UsageKey.from_string(peer_workflow.item_id)
+                ora_block = store.get_item(ora_block_key)
+
+                workflow_update_data = _add_ora_data(workflow_update_data=workflow_update_data,
+                                                     course_id=peer_workflow.course_id,
+                                                     item_id=peer_workflow.item_id,
+                                                     submission_uuid=peer_workflow.submission_uuid,
+                                                     assessment_requirements=ora_block.workflow_requirements())
```

Comment:
This still has redundancy, you never add the `item_id` to the cache, so this will look up the ORA multiple times. This whole flow seems a bit convoluted and can be better thought out.

Proposed:
```
get peer_workflows
    collected_data = {}
    for workflow in peer_workflow:
        if peer_workflow.course_id not in collected_data:
            get course_data
            collected_data.coursed.push(course_data)
       if peer_workflow.item_id not in collected_data
            get ora_data
            collected_data.courses[peer_workflow.course_id].assessments = ora_data
       # for all workflows
       add submission data
    return collected_data
```

Reply:
item_id is added when we add entire course object, line #379, I don't see redundancy, it is actually quite optimized adding ora data without additional lookup.  
BTW, your logic does not include flexible peer grading verification.